File: ЛР3.cpp

#include <iostream>
#include <vector>
#include <list>
#include <cmath>
#include <limits>
#include <algorithm>
#include <cassert>
#include <optional>
#include <unordered_map>

using namespace std;


// 1) БИНАРНАЯ КУЧА (Min-Heap)
class BinaryHeap {
private:
    // Внутренний массив для хранения элементов кучи
    vector<int> data;

    // Индексы в куче
    inline int parent(int i) const { return (i - 1) / 2; }
    inline int left(int i) const { return 2 * i + 1; }
    inline int right(int i) const { return 2 * i + 2; }

    // Восстановление свойства кучи снизу вверх (после вставки)
    void heapifyUp(int i) {
        while (i > 0 && data[parent(i)] > data[i]) {
            swap(data[i], data[parent(i)]);
            i = parent(i);
        }
    }

    // Восстановление свойства кучи сверху вниз (после удаления корня)
    void heapifyDown(int i) {
        int n = (int)data.size();
        while (true) {
            int l = left(i), r = right(i), smallest = i;
            if (l < n && data[l] < data[smallest]) smallest = l;
            if (r < n && data[r] < data[smallest]) smallest = r;
            if (smallest == i) break;
            swap(data[i], data[smallest]);
            i = smallest;
        }
    }

public:
    BinaryHeap() = default;

    // Вставка значения
    void insert(int val) {
        data.push_back(val);
        heapifyUp((int)data.size() - 1);
    }

    // Извлечение минимального элемента (корня). Возвращает optional<int>.
    optional<int> extractMin() {
        if (data.empty()) return nullopt;
        int res = data[0];
        data[0] = data.back();
        data.pop_back();
        if (!data.empty()) heapifyDown(0);
        return res;
    }

    bool empty() const { return data.empty(); }
    size_t size() const { return data.size(); }
};

// 2) БИНОМИАЛЬНАЯ КУЧА (Binomial Heap, min-heap)


struct BinNode {
    int key;
    int degree;            // степень (количество детей)
    BinNode* parent;
    BinNode* child;        // указатель на список детей (связан как однонаправленный sibling)
    BinNode* sibling;      // указатель на следующего брата в списке корней/детей

    BinNode(int k)
        : key(k), degree(0), parent(nullptr), child(nullptr), sibling(nullptr) {}
};

// Слияние двух биномиальных деревьев той же степени: возвращает корень результата
static BinNode* binomialLink(BinNode* a, BinNode* b) {
    // Пусть a.key <= b.key — делаем b дочерним у a
    b->parent = a;
    b->sibling = a->child;
    a->child = b;
    a->degree += 1;
    return a;
}

class BinomialHeap {
private:
    // Голова списка корней (сортированного по неубыванию степени)
    BinNode* head = nullptr;

    // Слияние двух списков корней (по возрастанию degree) — вспомогательная часть merge
    static BinNode* mergeRootLists(BinNode* h1, BinNode* h2) {
        if (!h1) return h2;
        if (!h2) return h1;
        BinNode* head = nullptr;
        BinNode* tail = nullptr;

        if (h1->degree <= h2->degree) {
            head = tail = h1; h1 = h1->sibling;
        } else {
            head = tail = h2; h2 = h2->sibling;
        }
        while (h1 && h2) {
            if (h1->degree <= h2->degree) {
                tail->sibling = h1; tail = h1; h1 = h1->sibling;
            } else {
                tail->sibling = h2; tail = h2; h2 = h2->sibling;
            }
        }
        tail->sibling = h1 ? h1 : h2;
        return head;
    }

public:
    BinomialHeap() = default;

    // Слияние (merge) с другой кучей — объединяет списки корней и корректно связывает деревья
    void merge(BinomialHeap& other) {
        BinNode* newHead = mergeRootLists(this->head, other.head);
        this->head = nullptr;
        other.head = nullptr;
        if (!newHead) return;

        BinNode* prev = nullptr;
        BinNode* curr = newHead;
        BinNode* next = curr->sibling;

        while (next) {
            if (curr->degree != next->degree ||
                (next->sibling && next->sibling->degree == curr->degree)) {
                // просто сдвигаем окно
                prev = curr;
                curr = next;
            } else {
                // curr.degree == next.degree и либо next->sibling.degree != curr.degree
                if (curr->key <= next->key) {
                    // присоединяем next к curr
                    curr->sibling = next->sibling;
                    binomialLink(curr, next);
                } else {
                    // присоединяем curr к next
                    if (prev) prev->sibling = next;
                    else newHead = next;
                    binomialLink(next, curr);
                    curr = next;
                }
            }
            next = curr->sibling;
        }

        this->head = newHead;
    }

    // Вставка одного ключа: создаём временную кучу из одного узла и сливаем
    void insert(int key) {
        BinomialHeap temp;
        temp.head = new BinNode(key);
        merge(temp);
    }

    // Поиск минимального ключа (без удаления)
    optional<int> getMin() const {
        if (!head) return nullopt;
        BinNode* cur = head;
        BinNode* best = cur;
        while (cur) {
            if (cur->key < best->key) best = cur;
            cur = cur->sibling;
        }
        return best ? optional<int>(best->key) : nullopt;
    }

    // Извлечение минимума: находим корень с минимальным ключом, удаляем его из корневого списка,
    // создаём новую кучу из его детей (обратный порядок) и сливаем.
    optional<int> extractMin() {
        if (!head) return nullopt;
        // найти минимальный корень и его предка
        BinNode* prevMin = nullptr;
        BinNode* minNode = head;
        BinNode* prev = nullptr;
        BinNode* cur = head;
        int minKey = cur->key;
        while (cur) {
            if (cur->key < minKey) {
                minKey = cur->key;
                prevMin = prev;
                minNode = cur;
            }
            prev = cur;
            cur = cur->sibling;
        }

        // удаляем minNode из списка корней
        if (!prevMin) head = minNode->sibling;
        else prevMin->sibling = minNode->sibling;

        // переворачиваем список детей minNode и формируем новую кучу
        BinNode* child = minNode->child;
        BinNode* rev = nullptr;
        while (child) {
            BinNode* next = child->sibling;
            child->sibling = rev;
            child->parent = nullptr;
            rev = child;
            child = next;
        }

        BinomialHeap temp;
        temp.head = rev;
        // освобождение minNode (для простоты: delete). В реальном коде стоит аккуратно всё деаллочить.
        delete minNode;

        // слияние текущей кучи и temp
        merge(temp);

        return optional<int>(minKey);
    }

    bool empty() const { return head == nullptr; }
};


// 3) КУЧА ФИБОНАЧИ (Fibonacci Heap, min-heap)


struct FibNode {
    int key;
    int degree;
    bool mark;
    FibNode* parent;
    FibNode* child;
    FibNode* left;
    FibNode* right;

    FibNode(int k)
        : key(k), degree(0), mark(false), parent(nullptr), child(nullptr) {
        left = right = this;
    }
};

class FibonacciHeap {
private:
    FibNode* minNode = nullptr;
    size_t n = 0;

    // Добавить узел x в корневой список (справа от minNode)
    void addToRootList(FibNode* x) {
        if (!minNode) {
            minNode = x;
            x->left = x->right = x;
        } else {
            x->right = minNode->right;
            x->left = minNode;
            minNode->right->left = x;
            minNode->right = x;
            if (x->key < minNode->key) minNode = x;
        }
    }

    // Удалить x из списка (корневого или детского)
    void removeFromList(FibNode* x) {
        x->left->right = x->right;
        x->right->left = x->left;
    }

    // Сделать y дочерним у x (используется в consolidate)
    void link(FibNode* y, FibNode* x) {
        // удалить y из корневого списка
        removeFromList(y);
        // присоединить к списку детей x
        y->parent = x;
        y->mark = false;
        if (!x->child) {
            x->child = y;
            y->left = y->right = y;
        } else {
            y->right = x->child->right;
            y->left = x->child;
            x->child->right->left = y;
            x->child->right = y;
        }
        x->degree++;
    }

    // Консолидация: объединение деревьев одинаковой степени
    void consolidate() {
        if (n == 0) return;
        // максимальная возможная степень <= floor(log2(n))
        size_t D = (size_t)(std::log2(max<size_t>(1, n)) + 2);
        vector<FibNode*> A(D, nullptr);

        // Соберём все корневые узлы в список (чтобы безопасно модифицировать)
        vector<FibNode*> roots;
        if (minNode) {
            FibNode* cur = minNode;
            do {
                roots.push_back(cur);
                cur = cur->right;
            } while (cur != minNode);
        }

        for (FibNode* w : roots) {
            FibNode* x = w;
            size_t d = x->degree;
            while (d >= A.size()) {
                A.resize(A.size() * 2, nullptr);
            }
            while (A[d]) {
                FibNode* y = A[d];
                if (x->key > y->key) swap(x, y);
                link(y, x);
                A[d] = nullptr;
                d++;
                if (d >= A.size()) A.resize(A.size() * 2, nullptr);
            }
            A[d] = x;
        }

        // восстановим корневой список и найдём новый minNode
        minNode = nullptr;
        for (FibNode* node : A) {
            if (!node) continue;
            // отрежем связи влево/вправо — сделаем каждый node отдельным циклом
            node->left = node->right = node;
            if (!minNode) {
                minNode = node;
            } else {
                // добавить node в корневой список
                node->right = minNode->right;
                node->left = minNode;
                minNode->right->left = node;
                minNode->right = node;
                if (node->key < minNode->key) minNode = node;
            }
        }
    }

public:
    FibonacciHeap() = default;

    // Вставка нового ключа в кучу
    FibNode* insert(int key) {
        FibNode* node = new FibNode(key);
        addToRootList(node);
        n++;
        return node;
    }

    // Извлечение минимума: возвращаем ключ или nullopt если пусто
    optional<int> extractMin() {
        FibNode* z = minNode;
        if (!z) return nullopt;

        // Перенести всех детей z в корневой список
        if (z->child) {
            // Соберём детей в вектор чтобы безопасно итерировать
            vector<FibNode*> children;
            FibNode* c = z->child;
            do {
                children.push_back(c);
                c = c->right;
            } while (c != z->child);

            for (FibNode* x : children) {
                // удаляем связь x <-> siblings (в рамках child list)
                removeFromList(x);
                // добавляем x в корневой список
                addToRootList(x);
                x->parent = nullptr;
            }
            z->child = nullptr;
        }

        // Удаляем z из корневого списка
        removeFromList(z);
        if (z == z->right) {
            minNode = nullptr;
        } else {
            minNode = z->right;
            consolidate();
        }

        int ret = z->key;
        delete z; // удалить узел (в реальном коде можно хранить и переиспользовать)
        n--;
        return optional<int>(ret);
    }

    bool empty() const { return minNode == nullptr; }
    size_t size() const { return n; }
};

// 4) ХЕШ-ТАБЛИЦА (цепочки) — простая реализация

template<typename K, typename V>
class HashTable {
private:
    vector<list<pair<K,V>>> table;
    size_t capacity;

    size_t indexFor(const K& key) const {
        // Используем std::hash и берем мод capacity
        return std::hash<K>{}(key) % capacity;
    }

public:
    explicit HashTable(size_t cap = 16) : table(cap), capacity(cap) {}

    // Вставка или обновление пары (key, value)
    void put(const K& key, const V& value) {
        size_t idx = indexFor(key);
        for (auto &p : table[idx]) {
            if (p.first == key) { p.second = value; return; }
        }
        table[idx].emplace_back(key, value);
    }

    // Получение значения по ключу, optional (nullopt если нет)
    optional<V> get(const K& key) const {
        size_t idx = indexFor(key);
        for (const auto &p : table[idx]) {
            if (p.first == key) return optional<V>(p.second);
        }
        return nullopt;
    }

    // Удаление по ключу, возвращает true если удалили
    bool remove(const K& key) {
        size_t idx = indexFor(key);
        auto &lst = table[idx];
        for (auto it = lst.begin(); it != lst.end(); ++it) {
            if (it->first == key) {
                lst.erase(it);
                return true;
            }
        }
        return false;
    }
};

File: ЛР3.py

# 1. БИНАРНАЯ КУЧА (Binary Heap)


class BinaryHeap:
    """Простая реализация минимальной бинарной кучи."""

    def __init__(self):
        self.data = []  # внутренний массив кучи

    def _parent(self, i): return (i - 1) // 2
    def _left(self, i): return 2 * i + 1
    def _right(self, i): return 2 * i + 2

    def _heapify_up(self, i):
        """Восстановление свойства кучи при вставке элемента вверх."""
        while i != 0 and self.data[self._parent(i)] > self.data[i]:
            # Меняем местами с родителем
            self.data[i], self.data[self._parent(i)] = self.data[self._parent(i)], self.data[i]
            i = self._parent(i)

    def _heapify_down(self, i):
        """Восстановление свойства кучи при удалении, спуская вниз."""
        smallest = i
        l, r = self._left(i), self._right(i)

        # Проверяем левый ребёнок
        if l < len(self.data) and self.data[l] < self.data[smallest]:
            smallest = l
        # Проверяем правый ребёнок
        if r < len(self.data) and self.data[r] < self.data[smallest]:
            smallest = r

        if smallest != i:
            # Меняем элементы
            self.data[i], self.data[smallest] = self.data[smallest], self.data[i]
            self._heapify_down(smallest)

    def insert(self, value):
        """Вставка нового значения."""
        self.data.append(value)
        self._heapify_up(len(self.data) - 1)

    def extract_min(self):
        """Удаление и возврат минимального элемента."""
        if not self.data:
            return None
        root = self.data[0]
        self.data[0] = self.data[-1]
        self.data.pop()
        if self.data:
            self._heapify_down(0)
        return root

# 2. БИНОМИАЛЬНАЯ КУЧА (Binomial Heap)

class BinomialNode:
    """Узел биномиальной кучи."""
    def __init__(self, key):
        self.key = key
        self.degree = 0
        self.parent = None
        self.child = None
        self.sibling = None


def merge_trees(t1, t2):
    """Слияние двух биномиальных деревьев одинаковой степени."""
    if t1.key > t2.key:
        t1, t2 = t2, t1  # t1 должен быть корнем с меньшим ключом
    t2.parent = t1
    t2.sibling = t1.child
    t1.child = t2
    t1.degree += 1
    return t1


class BinomialHeap:
    """Реализация биномиальной кучи (минимальной)."""

    def __init__(self):
        self.head = None

    def _merge_roots(self, h1, h2):
        """Слияние списков корней двух куч по степени."""
        if not h1:
            return h2
        if not h2:
            return h1

        if h1.degree <= h2.degree:
            head = h1
            h1 = h1.sibling
        else:
            head = h2
            h2 = h2.sibling

        tail = head

        while h1 and h2:
            if h1.degree <= h2.degree:
                tail.sibling = h1
                h1 = h1.sibling
            else:
                tail.sibling = h2
                h2 = h2.sibling
            tail = tail.sibling

        tail.sibling = h1 if h1 else h2
        return head

    def merge(self, other):
        """Слияние с другой биномиальной кучей."""
        new_head = self._merge_roots(self.head, other.head)
        if not new_head:
            return

        prev = None
        curr = new_head
        next = curr.sibling

        while next:
            # Условия: либо степени разные, либо 3 одинаковых подряд
            if curr.degree != next.degree or (next.sibling and next.sibling.degree == curr.degree):
                prev = curr
                curr = next
            else:
                if curr.key <= next.key:
                    curr.sibling = next.sibling
                    merge_trees(curr, next)
                else:
                    if prev:
                        prev.sibling = next
                    else:
                        new_head = next
                    merge_trees(next, curr)
                    curr = next
            next = curr.sibling

        self.head = new_head

    def insert(self, key):
        """Вставка нового элемента."""
        temp = BinomialHeap()
        temp.head = BinomialNode(key)
        self.merge(temp)

    def get_min(self):
        """Возврат минимального ключа без удаления."""
        if not self.head:
            return None
        x = self.head
        mn = x
        while x:
            if x.key < mn.key:
                mn = x
            x = x.sibling
        return mn.key


# 3. КУЧА ФИБОНАЧЧИ (Fibonacci Heap)

import math

class FibNode:
    """Узел кучи Фибоначчи."""
    def __init__(self, key):
        self.key = key
        self.degree = 0
        self.mark = False
        self.parent = None
        self.child = None
        self.left = self
        self.right = self


class FibonacciHeap:
    """Реализация минимальной кучи Фибоначчи."""

    def __init__(self):
        self.min = None  # указатель на минимальный узел
        self.n = 0       # количество элементов

    def insert(self, key):
        """Вставка нового элемента."""
        node = FibNode(key)
        if not self.min:
            self.min = node
        else:
            # Вставка в корневой список
            node.right = self.min.right
            node.left = self.min
            self.min.right.left = node
            self.min.right = node
            if node.key < self.min.key:
                self.min = node
        self.n += 1
        return node

    def _link(self, y, x):
        """Присоединение дерева y к узлу x (увеличение степени x)."""
        # Удаляем y из корневого списка
        y.left.right = y.right
        y.right.left = y.left

        y.parent = x
        if not x.child:
            x.child = y
            y.right = y.left = y
        else:
            y.right = x.child.right
            y.left = x.child
            x.child.right.left = y
            x.child.right = y

        x.degree += 1
        y.mark = False

    def extract_min(self):
        """Удаление и возврат минимального элемента."""
        z = self.min
        if z:
            # Переносим детей z в корневой список
            if z.child:
                children = []
                c = z.child
                while True:
                    children.append(c)
                    c = c.right
                    if c == z.child:
                        break

                for x in children:
                    # Отсоединяем от ребёнка
                    x.left.right = x.right
                    x.right.left = x.left

                    # Добавляем в корневой список
                    x.left = self.min
                    x.right = self.min.right
                    self.min.right.left = x
                    self.min.right = x
                    x.parent = None

            # Удаляем z из корневого списка
            z.left.right = z.right
            z.right.left = z.left

            if z == z.right:
                self.min = None
            else:
                self.min = z.right
                self._consolidate()
            self.n -= 1

        return z.key if z else None

    def _consolidate(self):
        """Фаза консолидации после удаления минимального элемента."""
        A = [None] * int(math.log(self.n, 2) + 3)
        nodes = []
        x = self.min

        # Собираем узлы корневого списка
        while True:
            nodes.append(x)
            x = x.right
            if x == self.min:
                break

        for w in nodes:
            x = w
            d = x.degree
            while A[d]:
                y = A[d]
                if y.key < x.key:
                    x, y = y, x
                self._link(y, x)
                A[d] = None
                d += 1
            A[d] = x

        # Поиск нового минимума
        self.min = None
        for i in A:
            if i:
                if not self.min or i.key < self.min.key:
                    self.min = i


# 4. ХЕШ-Таблица (Hash Table) с цепочками


class HashTable:
    """Хеш-таблица на основе цепочек (списки для коллизий)."""

    def __init__(self, size=16):
        self.size = size
        self.table = [[] for _ in range(size)]

    def _hash(self, key):
        """Хеш-функция: Python hash мод размер."""
        return hash(key) % self.size

    def put(self, key, value):
        """Сохранение значения по ключу."""
        idx = self._hash(key)
        for i, (k, v) in enumerate(self.table[idx]):
            if k == key:
                # Обновляем существующий ключ
                self.table[idx][i] = (key, value)
                return
        # Иначе добавляем новую пару
        self.table[idx].append((key, value))

    def get(self, key):
        """Получение значения по ключу."""
        idx = self._hash(key)
        for k, v in self.table[idx]:
            if k == key:
                return v
        return None

    def remove(self, key):
        """Удаление элемента."""
        idx = self._hash(key)
        for i, (k, v) in enumerate(self.table[idx]):
            if k == key:
                del self.table[idx][i]
                return True
        return False



